packages/truffle-decode-utils/src/types/values.ts

```diff
@@ -0,0 +1,1423 @@
+import debugModule from "debug";
+const debug = debugModule("decode-utils:values");
+
+//objects for Solidity values
+
+//Note: This is NOT intended to represent every possible value that exists
+//in Solidity!  Only possible values of variables.  (Though there may be
+//some expansion in the future.)  We do however count the builtin variables
+//msg, block, and tx as variables (not other builtins though for now) so
+//there is some support for the magic type.
+
+//We don't include fixed and ufixed for now.  Those will be added when
+//implemented.
+
+//NOTE: not all of these optional fields are actually implemented. Some are
+//just intended for the future.  More optional fields may be added in the
+//future.
+
+//Note: Many of the errors defined here deliberately *don't* extend Error.
+//This is because their intended use is a little different.  Only the ones
+//that are for throwing extend Error.
+
+import BN from "bn.js";
+import { Types } from "./types";
+import util from "util";
+import { AstDefinition } from "../ast";
+import { Definition as DefinitionUtils } from "../definition";
+
+export namespace Values {
+
+  /*
+   * SECTION 0: Some irrelevant stuff for dealing with util.inspect.custom.
+   * You can probably ignore these.
+   */
+
+  //we'll need to write a typing for the options type ourself, it seems; just
+  //going to include the relevant properties here
+  interface InspectOptions {
+    stylize?: (toMaybeColor: string, style?: string) => string;
+    colors: boolean;
+    breakLength: number;
+  }
+
+  //HACK -- inspect options are ridiculous, I swear >_>
+  function cleanStylize(options: InspectOptions) {
+    return Object.assign({}, ...Object.entries(options).map(
+      ([key,value]) =>
+        key === "stylize"
+          ? {}
+          : {[key]: value}
+    ));
+  }
+
+  /*
+   * SECTION 1: Generic types for values in general (including errors).
+   */
```

Comment:
I really appreciate the care towards organizing sections of code so that this file can be read sequentially. Literate programming :)

I know I've been on about TypeDoc, t should be relatively little effort getting to get these doc strings into HTML. I don't know how to preserve order with TypeDoc, but I bet there's a way or a hack. It feels a waste to hide this knowledge behind the daunting wall of a 1,400 line file.

Reply:
Hm, worth looking into, I guess?